zson/formatter.go

```diff
@@ -0,0 +1,401 @@
+package zson
+
+import (
+	"errors"
+	"fmt"
+	"strconv"
+	"strings"
+
+	"github.com/brimsec/zq/zcode"
+	"github.com/brimsec/zq/zng"
+)
+
+type Formatter struct {
+	typedefs    typemap
+	tab         int
+	newline     string
+	whitespace  string
+	typeTab     int
+	typeNewline string
+	nid         int
+	types       *TypeTable
+	builder     strings.Builder
+	stack       []strings.Builder
+}
+
+func NewFormatter(pretty int) *Formatter {
+	newline := ""
+	if pretty > 0 {
+		newline = "\n"
+	}
+	return &Formatter{
+		typedefs:   make(typemap),
+		tab:        pretty,
+		newline:    newline,
+		whitespace: strings.Repeat(" ", 80),
+	}
+}
+
+func (f *Formatter) push() {
+	f.stack = append(f.stack, f.builder)
+	f.builder = strings.Builder{}
+}
+
+func (f *Formatter) pop() {
+	n := len(f.stack)
+	f.builder = f.stack[n-1]
+	f.stack = f.stack[:n-1]
+}
+
+func (f *Formatter) FormatRecord(rec *zng.Record) (string, error) {
+	f.builder.Reset()
+	if err := f.formatValueAndDecorate(rec.Alias, rec.Raw); err != nil {
+		return "", err
+	}
+	return f.builder.String(), nil
+}
+
+func (f *Formatter) formatValueAndDecorate(typ zng.Type, bytes zcode.Bytes) error {
+	known := f.typedefs.exists(typ)
+	if err := f.formatValue(0, typ, bytes, known, false); err != nil {
+		return err
+	}
+	f.decorate(typ, false, bytes == nil)
+	return nil
+}
+
+func (f *Formatter) formatValue(indent int, typ zng.Type, bytes zcode.Bytes, parentKnown, decorate bool) error {
+	known := parentKnown || f.typedefs.exists(typ)
+	if bytes == nil {
+		f.build("null")
+		f.decorate(typ, parentKnown, true)
+		return nil
+	}
+	var err error
+	var null bool
+	switch t := typ.(type) {
+	default:
+		f.build(typ.ZSONOf(bytes))
+	case *zng.TypeAlias:
+		if err := f.formatValue(indent, t.Type, bytes, known, false); err != nil {
+			return err
+		}
```

Comment:
No need to check err here.
```suggestion
		err = f.formatValue(indent, t.Type, bytes, known, false)
```